Get more specific HttpContent MethodInfos

campersau opened the pull request:
## Bug

Fixes: NuGet/Home#8661

dotnet core includes a [new](https://github.com/dotnet/corefx/pull/27029) [`SerializeToStreamAsync`](https://github.com/dotnet/corefx/blob/a528006b4536e5b73227285b57d2a56116532506/src/System.Net.Http/src/System/Net/Http/HttpContent.cs#L317-L318) overload which is currently internal, but can be received via reflection. Which is done in the `HttpContentWrapper` introduced in https://github.com/NuGet/NuGet.Client/pull/3078 . Since the `GetMethod` wasn't very specifc about the method parameters this throws a `TypeInitializerException` with an inner `AmbiguousMatchException` on dotnet core 3.0.

Fixes: https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/issues/866
Regression: Yes
* Last working version: before https://github.com/NuGet/NuGet.Client/pull/3078
* How are we preventing it in future: running tests on dotnet core 3.x

## Fix

Get more specific MethodInfo.

## Testing/Validation

Tests Added: No
Reason for not adding tests: covered by existing tests
Validation:

campersau committed:
Don't reuse HttpRequestMessage objects

zivkan reviewed (changes requested):
Now I regret approving and merging the previous PR. I somehow didn't notice it used reflection. I think it would be better to remove the use of reflection, even if it requires our code to be a few lines longer.

What benefit does using `SerializeToStreamAsync` bring over using APIs that give us compile-time guarantees?

campersau commented:
For `SerializeToStreamAsync` we could use `_httpContent.CopyToAsync(stream, context);` instead which does the [same thing](https://github.com/dotnet/corefx/blob/a528006b4536e5b73227285b57d2a56116532506/src/System.Net.Http/src/System/Net/Http/HttpContent.cs#L349).
But for `TryComputeLength` there is no public method alternative.

zivkan commented:
Does `TryComputeLength` ever get used? Is there any problem if we just return false?

You could also consider using `#if IS_DESKTOP` (check the actual name we use, but it's something like that) to do different things on .NET Framework and then use `#else` for the .NET Core codepath, where you might be able to use the new APIs, assuming they're in the .NET Standard 2.0 version of System.Net.Http.

If these APIs are .NET Core specific, how does this even work on .NET Framework?

campersau commented:
Yes, it is used [here](https://github.com/dotnet/corefx/blob/d3911035f2ba3eb5c44310342cc1d654e42aa316/src/System.Net.Http/src/System/Net/Http/Headers/HttpContentHeaders.cs#L82) to set the `Content-Length` header.

Not sure I understand what you mean with .NET Framework and .NET Core. Both of these APIs exist in both of them. But in .NET Core some more internal APIs were added (which aren't used here) but interferred with the reflection code.

zivkan commented:
what I mean about .NET Framework and .NET Core is if you need to use different APIs when compiling against the different TFMs, use 
```csharp
#if IS_DESKTOP
  // use Windows-only, .NET Framework specific APIs here
#else
  // use .NET Standard APIs here
#endif
```

I really don't feel comfortable about using reflection to access internal APIs. Those APIs being internal mean that they could change at any release, making NuGet fragile. I only want to use public, supported APIs.

About `TryComputeLength`, I meant does NuGet use it? `HttpContentWrapper` is an internal class in the NuGet assembly, so it only matters if one of our own codepaths will call it. Given the wrapper is about wrapping outgoing requests, not incoming responses, what's the harm if we don't have a `Content-Length` header? The only requests I can think of that will have a body is pushing nupkgs. Can push still work if we don't have a `Content-Length` header?

Perhaps a better option is to add a nullable int to `HttpContentWrapper`'s constructor, and `TryComputeLength` just returns that value. Whatever constructs `HttpContentWrapper` is then responsible for calling `TryComputeLength` on the original `HttpContent` once, and pass the value to the wrapper constructor each time. This way we can avoid the fragility of using reflection.

campersau commented:
I don't need to call different APIs on different TFMs. The only problem here was that on dotnet core there is a new internal overload which interferred with the previous reflection lookup logic.

I don't call any `internal` APIs but `protected` ones, thats why reflection is used.

`TryComputeLength` isn't used in `NuGet.Client` directly but in other `HttpContent` implementations which `NuGet.Client` is using.
I have added another testcase which demonstratest that `TryComputeLength` is being called.

I am not sure if packages can be pushed without `Content-Length` header, but currently a `Content-Length` header is set.

zivkan commented:
I briefly discussed this with my team mates today, and we'd prefer a solution that doesn't rely on reflection if possible. Using a protected method via reflection is much better than using an internal method, but we're still worried about the maintainability: risk of breaking if someone refactors the code, risk of breaking if we target different TFMs without compile time type checking.

So, we don't want to rush accepting this PR, but the problems it causes (completely breaking in .NET Core 3) is worse than reverting the change (wrong headers in a small subset of retries?), so we reverted the first PR. Let's work on a good long term solution.

campersau commented:
I have updated the PR to not use reflection for the `HttpContent` wrapper and added some tests for some common `HttpContent` implementations.

Note that the `System.Net.Http.WinHttpHandler` will be removed in the upcoming dotnet core version 5 (https://github.com/dotnet/runtime/pull/397) so the workaround with using it instead will no longer work (https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/pull/841).

donnie-msft commented:
I'm looking into this PR and getting familiar with the changes and potential impact.
I've rebased and gotten a good CI build from the changes as they are.

What I'm trying to figure out as time permits this week, is am I comfortable with any impacts on commands like `push` ? Are there any tests that can be added, or a manual test using a tool or custom Server, to prove that those commands' behavior is as expected when the HTTP retry occurs?

campersau commented:
I think the following test does something like that, it pushes a package and fails to authenticates a few times before succeeding (retries).
https://github.com/NuGet/NuGet.Client/blob/ecbb8a0f0d3ec28fbd037544470759d628873efc/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs#L696-L774

zivkan reviewed (approved):
Donnie did some manual testing, so I'm satisfied that this is unlikely to cause further issues.

donnie-msft reviewed (approved):
I've tested on a custom server where I set it up to return ServiceUnavailable for every other request. With your branch's nuget.exe, and I verified it still pushed the entire nupkg to the server, not corrupted.

> Pushing NuGet.LibraryModel.5.7.0-zlocal.34008.nupkg to 'http://localhost:50561'...
>   PUT http://localhost:50561/api/v2/package/
>   ServiceUnavailable http://localhost:50561/api/v2/package/ 39ms
>   PUT http://localhost:50561/api/v2/package/
>   Created http://localhost:50561/api/v2/package/ 167ms
> Your package was pushed.

So, approving and will get this merged.
